src/uct/tcp/sockcm/sockcm_ep.c

```diff
@@ -44,9 +47,173 @@ void uct_sockcm_ep_put_sock_id(uct_sockcm_ctx_t *sock_id_ctx)
     ucs_free(sock_id_ctx);
 }
 
+ucs_status_t uct_sockcm_ep_send_client_info(uct_sockcm_iface_t *iface, uct_sockcm_ep_t *ep)
+{
+    ssize_t transfer_len = 0;
+    ssize_t sent_len     = 0;
+    ssize_t sent_off     = 0;
+    uct_md_h dummy_md    = 0;
+    int retries          = 0;
+    ucs_status_t status;
+    unsigned num_resources;
+    uct_tl_device_resource_t *devices;
+    uct_sockcm_conn_param_t conn_param;
+
+    memset(&conn_param, 0, sizeof(uct_sockcm_conn_param_t));
+
+    status = uct_tcp_query_devices(dummy_md, &devices, &num_resources);
+    if (UCS_OK != status || num_resources == 0) {
+        ucs_error("sockcm unable to find a tcp-capable device");
+        status = (status == UCS_OK) ? UCS_ERR_IO_ERROR : status;
+        goto out;
+    }
+    ucs_debug("using device %s at client side\n", devices[0].name);
+
+    /* pack worker address into private data */
+    conn_param.length = ep->pack_cb(ep->pack_cb_arg, devices[0].name,
```

Comment:
assuming the connection establishment will always use the first device on the system??

Reply:
i think it should be:
1. check if UCX_TCP_IF_INCLUDE global conf set with desired device name and use it if so
1. if [1] is not set - get hostname from gethostname() and use it

Reply:
IMO if i/f is not set - detect i/f name from hostname .

Reply:
@Akshay-Venkatesh need to use the device which routes to the remote address provided when creating the client endpoint

Reply:
@yosefe One solution I can think of here is to iterate through the tcp devices returned and attempt `pack_cb` against every such device until success. My expectation is that one of the `tcp` tl resource devices will match against the list of devices returned by `uct_tcp_query_devices` and we break when there's a success. I was thinking along the following lines:
```diff
diff --git a/src/uct/tcp/sockcm/sockcm_ep.c b/src/uct/tcp/sockcm/sockcm_ep.c
index b96f9fd..c4c6715 100644
--- a/src/uct/tcp/sockcm/sockcm_ep.c
+++ b/src/uct/tcp/sockcm/sockcm_ep.c
@@ -54,6 +54,7 @@ ucs_status_t uct_sockcm_ep_send_client_info(uct_sockcm_iface_t *iface, uct_sockc
     uct_md_h dummy_md    = 0;
     ucs_status_t status;
     unsigned num_resources;
+    unsigned i;
     uct_tl_device_resource_t *devices;
     uct_sockcm_conn_param_t conn_param;
 
@@ -65,18 +66,24 @@ ucs_status_t uct_sockcm_ep_send_client_info(uct_sockcm_iface_t *iface, uct_sockc
         status = (status == UCS_OK) ? UCS_ERR_IO_ERROR : status;
         goto out;
     }
-    ucs_debug("using device %s at client side", devices[0].name);
 
+    for (i = 0; i < num_resources; i++) {
     /* pack worker address into private data */
-    conn_param.length = ep->pack_cb(ep->pack_cb_arg, devices[0].name,
-                                    (void*)conn_param.private_data);
-    if (conn_param.length < 0) {
-        ucs_error("sockcm client (iface=%p, ep = %p) failed to fill "
-                  "private data. status: %s",
-                  iface, ep, ucs_status_string(conn_param.length));
-        status = UCS_ERR_IO_ERROR;
-        goto out;
+        conn_param.length = ep->pack_cb(ep->pack_cb_arg, devices[i].name,
+                                        (void*)conn_param.private_data);
+        if (conn_param.length < 0) {
+            if (i == (num_resources - 1)) {
+                ucs_error("sockcm client (iface=%p, ep = %p) failed to fill "
+                          "private data. status: %s",
+                          iface, ep, ucs_status_string(conn_param.length));
+                status = UCS_ERR_IO_ERROR;
+                goto out;
+            }
+            continue;
+        }
+        break;
     }
+    ucs_debug("using device %s at client side", devices[i].name);
```
The main assumption I make in the code above is that there is some tl resource whose device name matches with the names returned by `uct_tcp_query_devices`. That said, I suppose `rdmacm` also makes this assumption in a way because it generates a device name such as `mlx5_0:1` from `cm_id` and expects a tl resource in aux tls to have a device which matches that. Please correct me if my understanding is wrong here.